experimental/Examples/NFT.idr

```diff
@@ -0,0 +1,230 @@
+TokenId : Type
+TokenId = Nat
+
+||| Account contains the balance (number of tokens owned)
+||| and a map of operator approvals of an associated address.
+record Account where
+  constructor MkAccount
+  ownedTokens : Nat
+  opApprovals : SortedMap Address Bool --operator approvals
+
+||| Token contains the info of a specific token
+record Token where
+  constructor MkToken
+  tokenOwner : Address
+  approved : Maybe Address -- approved address (there can at most be one address)
+
+||| The storage has type Storage which is a record with accounts and tokens.
+record Storage where
+    constructor MkStorage
+    accounts : SortedMap Address Account
+    tokens : SortedMap TokenId Token
+    version : Nat -- version of the token standard
+    totalSup : Nat
+    owner : Address -- owner of this NFT contract
+
+total emptyStorage : Storage
+emptyStorage =
+  MkStorage
+    empty
+    empty
+    1
+    0
+    "qwer"
+
+||| getAccount returns the account of an associated key hash.
+||| @address the key hash of the owner of the account
+total getAccount :
+(address : Address) -> SortedMap Address Account -> Account
+getAccount address accounts = case lookup address accounts of
+                      Nothing => MkAccount 0 empty
+                      (Just account) => account
+
+||| getAccountBal returns the balance of an associated key hash.
+||| @address the key hash of the owner of the account
+total getAccountBal : (address : Address) -> SortedMap Address Account -> Nat
+getAccountBal address accounts = case lookup address accounts of
+                      Nothing => 0
+                      (Just account) => ownedTokens account
+
+||| mint when the caller is the owner of the token contract,
+||| they can mint/add new tokens to the total supply
+||| @token the token to be added to the total supply
+||| @dest the address the new tokens to be added to
+||| @storage the original storage
+total mint :
+(token : TokenId) -> (dest : Address) -> (storage : Storage) -> Either Error Storage
+mint token dest storage =
+  let currentAcc = accounts storage
+      currentTok = tokens storage in
+    let destAcc = getAccount dest currentAcc in
+      case lookup token currentTok of
+        Just something => Left TokenAlreadyMinted
+        Nothing =>
+          case currentCaller == owner storage of
+            False => Left FailedToAuthenticate
+            True =>
+              let newAcc = insert dest (record {ownedTokens $= (+ 1)} destAcc) currentAcc
+                  newToken = insert token (MkToken dest Nothing) currentTok in
+                    Right
+                      (record
+                        {accounts = newAcc,
+                         tokens = newToken,
+                         totalSup $= (+ 1)
+                        } storage
+                      )
+
+total initStorage : Storage
+initStorage =
+  case mint 1 "qwer" emptyStorage of
+    Left _ => emptyStorage
+    Right s => s
+
+total storage : Storage
+storage =
```

Comment:
This global seems to be used in a lot of places where it seems like it should be a parameter, can we just get rid of it?

Reply:
Yes, I can get rid of it.